File: Daily Resources/Day07/DSA - Sliding Window Solutions/minimum_window_substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) return "";
        
        unordered_map<char, int> t_count, window_count;
        for (char c : t) t_count[c]++;
        
        int have = 0, need = t_count.size();
        int left = 0, min_len = INT_MAX, res_left = 0;
        
        for (int right = 0; right < s.size(); ++right) {
            char c = s[right];
            window_count[c]++;
            
            if (t_count.count(c) && window_count[c] == t_count[c]) {
                have++;
            }
            
            while (have == need) {
                if ((right - left + 1) < min_len) {
                    min_len = right - left + 1;
                    res_left = left;
                }
                
                window_count[s[left]]--;
                if (t_count.count(s[left]) && window_count[s[left]] < t_count[s[left]]) {
                    have--;
                }
                left++;
            }
        }
        
        return min_len == INT_MAX ? "" : s.substr(res_left, min_len);
    }
};

File: Daily Resources/Day07/DSA - Sliding Window Solutions/minimum_window_substring.py
class Solution:
    def minWindow(self, s, t):
        if not t or not s:
            return ""
        
        t_count = Counter(t)
        window_count = {}
        
        have, need = 0, len(t_count)
        res = [-1, -1]
        res_len = float("inf")
        left = 0
        
        for right in range(len(s)):
            c = s[right]
            window_count[c] = window_count.get(c, 0) + 1
            
            if c in t_count and window_count[c] == t_count[c]:
                have += 1
            
            while have == need:
                # Update result if smaller window is found
                if (right - left + 1) < res_len:
                    res = [left, right]
                    res_len = right - left + 1
                
                # Shrink window from left
                window_count[s[left]] -= 1
                if s[left] in t_count and window_count[s[left]] < t_count[s[left]]:
                    have -= 1
                left += 1
        
        l, r = res
        return s[l:r+1] if res_len != float("inf") else ""
